File: string_crypter.rb
# string_crypter.rb
require 'securerandom'
require 'base64'

STATIC_KEY_RB = "test_key"
SALT_LENGTH_BYTES_RB = 16 # 16 bytes = 32 hex characters

def generate_salt_rb
  SecureRandom.hex(SALT_LENGTH_BYTES_RB)
end

def xor_strings_rb(s1_bytes, s2_bytes)
  return s1_bytes if s2_bytes.empty?
  s1_bytes.zip(s2_bytes.cycle).map { |b1, b2| b1 ^ b2 }
end

def encrypt_rb(plaintext, key_string)
  salt_hex = generate_salt_rb
  data_with_salt_string = salt_hex + plaintext

  data_with_salt_bytes = data_with_salt_string.bytes
  key_bytes = key_string.bytes

  xored_bytes_array = xor_strings_rb(data_with_salt_bytes, key_bytes)
  # Convert array of byte values back to a packed byte string
  xored_packed_bytes = xored_bytes_array.pack('C*')
  Base64.strict_encode64(xored_packed_bytes)
end

def decrypt_rb(encrypted_data, key_string)
  begin
    decoded_packed_bytes = Base64.strict_decode64(encrypted_data)
    decoded_bytes = decoded_packed_bytes.bytes # Get array of byte values
    key_bytes = key_string.bytes

    decrypted_with_salt_bytes_array = xor_strings_rb(decoded_bytes, key_bytes)
    # Convert array of byte values back to a packed byte string, then to UTF-8 string
    decrypted_with_salt_string = decrypted_with_salt_bytes_array.pack('C*').force_encoding('UTF-8')

    salt_hex_len = SALT_LENGTH_BYTES_RB * 2
    if decrypted_with_salt_string.bytesize < salt_hex_len
      puts "Decryption error: Decrypted data is too short to contain salt."
      return nil
    end
    # Substring based on character count, assuming salt is ASCII/UTF-8 compatible hex
    decrypted_with_salt_string[salt_hex_len..-1]
  rescue ArgumentError => e # Catches invalid base64 string
    puts "Decryption error: Invalid Base64 string. #{e.message}"
    nil
  rescue StandardError => e
    puts "Decryption error: #{e.message}"
    nil
  end
end

def process_string_rb(type, data)
  key = STATIC_KEY_RB
  case type
  when 'e'
    encrypt_rb(data, key)
  when 'd'
    decrypt_rb(data, key)
  else
    puts "Error: Invalid type specified. Use 'e' for encrypt or 'd' for decrypt."
    nil
  end
end

# Main test area
if __FILE__ == $0
  puts "--- Ruby String Encryption/Decryption Tests ---"

  original_text = "Hello from Ruby!"
  puts "Original Text: #{original_text}"

  encrypted_text = process_string_rb('e', original_text)
  if encrypted_text
    puts "Encrypted (Ruby): #{encrypted_text}"

    decrypted_text = process_string_rb('d', encrypted_text)
    if decrypted_text
      puts "Decrypted (Ruby): #{decrypted_text}"
      if decrypted_text == original_text
        puts "Ruby Encryption/Decryption Test: SUCCESSFUL"
      else
        puts "Ruby Encryption/Decryption Test: FAILED"
        puts "Expected: #{original_text}, Got: #{decrypted_text}"
      end
    else
      puts "Ruby Decryption FAILED (result was nil)."
    end
  else
    puts "Ruby Encryption FAILED (result was nil)."
  end

  puts "\n--- Interoperability Test (Ruby decrypts Python) ---"
  # Example: String "Hello from Python for Ruby!" encrypted by Python
  python_encrypted = "NzM2MzMxMzEzNjY0NjEzMDYxMzQzNjY0NjIzNjYyMzQzNFN1Z2RjY2RjZWNkZWNlY2RjZWRlY2VjZGNlZGVjZWNkY2VkZWNlY2RjZWRlY2VjZGNlZGVjZWNkY2VkZWNlY2Q=" # Placeholder
  expected_python_decryption = "Hello from Python for Ruby!" # Placeholder

  puts "Python Encrypted: #{python_encrypted}"
  decrypted_from_python = process_string_rb('d', python_encrypted)
  if decrypted_from_python
    puts "Decrypted by Ruby: #{decrypted_from_python}"
    if decrypted_from_python == expected_python_decryption
      puts "Ruby decryption of Python string: SUCCESSFUL (with correct Python output)"
    else
      puts "Ruby decryption of Python string: FAILED or placeholder data used."
      puts "Expected: #{expected_python_decryption}, Got: #{decrypted_from_python}"
    end
  else
    puts "Ruby decryption of Python string FAILED (result was nil)."
  end
  puts "Note: For the Python interop test to be meaningful, replace 'python_encrypted' with actual output from your Python script for the string '#{expected_python_decryption}'."


  puts "\n--- Interoperability Test (Ruby encrypts for others) ---"
  ruby_msg_for_others = "Hello from Ruby for other languages!"
  ruby_encrypted_for_others = process_string_rb('e', ruby_msg_for_others)
  if ruby_encrypted_for_others
    puts "Ruby Encrypted for others (Original: '#{ruby_msg_for_others}'): #{ruby_encrypted_for_others}"
    puts "Take this string and try to decrypt it using process_string('d', ...) in other languages."
  else
    puts "Ruby encryption for others FAILED (result was nil)."
  end
end

File: string_crypter.dart
// string_crypter.dart
import 'dart:convert'; // For base64 and utf8
import 'dart:math';   // For Random
import 'dart:typed_data'; // For Uint8List

const String staticKeyDart = "test_key";
const int saltLengthBytesDart = 16; // 16 bytes = 32 hex characters

String _bytesToHexString(Uint8List bytes) {
  return bytes.map((byte) => byte.toRadixString(16).padLeft(2, '0')).join('');
}

String generateSaltDart() {
  final random = Random.secure();
  final saltBytes = Uint8List(saltLengthBytesDart);
  for (int i = 0; i < saltLengthBytesDart; i++) {
    saltBytes[i] = random.nextInt(256);
  }
  return _bytesToHexString(saltBytes);
}

Uint8List xorByteArraysDart(Uint8List input, Uint8List key) {
  final output = Uint8List(input.length);
  if (key.isEmpty) return Uint8List.fromList(input); // Return a copy if key is empty
  for (int i = 0; i < input.length; i++) {
    output[i] = input[i] ^ key[i % key.length];
  }
  return output;
}

String encryptDart(String plaintext, String keyString) {
  final saltHex = generateSaltDart();
  final dataWithSaltString = saltHex + plaintext;

  final dataWithSaltBytes = utf8.encode(dataWithSaltString);
  final keyBytes = utf8.encode(keyString);

  final xoredBytes = xorByteArraysDart(Uint8List.fromList(dataWithSaltBytes), Uint8List.fromList(keyBytes));
  return base64.encode(xoredBytes);
}

String? decryptDart(String encryptedData, String keyString) {
  try {
    final decodedBytes = base64.decode(encryptedData);
    final keyBytes = utf8.encode(keyString);

    final decryptedWithSaltBytes = xorByteArraysDart(decodedBytes, Uint8List.fromList(keyBytes));
    // Convert the entire byte array to string first
    final decryptedWithSaltString = utf8.decode(decryptedWithSaltBytes);

    final saltHexLen = saltLengthBytesDart * 2;
    if (decryptedWithSaltString.length < saltHexLen) {
      print("Decryption error: Decrypted data is too short to contain salt.");
      return null;
    }
    // Then extract the substring after the salt
    return decryptedWithSaltString.substring(saltHexLen);
  } catch (e) {
    print("Decryption error: ${e.toString()}");
    return null;
  }
}

String? processStringDart(String type, String data) {
  final key = staticKeyDart;
  switch (type) {
    case 'e':
      return encryptDart(data, key);
    case 'd':
      return decryptDart(data, key);
    default:
      print("Error: Invalid type specified. Use 'e' for encrypt or 'd' for decrypt.");
      return null;
  }
}

void main() {
  print("--- Dart String Encryption/Decryption Tests ---");

  final originalText = "Hello from Dart!";
  print("Original Text: $originalText");

  final encryptedText = processStringDart('e', originalText);
  if (encryptedText != null) {
    print("Encrypted (Dart): $encryptedText");

    final decryptedText = processStringDart('d', encryptedText);
    if (decryptedText != null) {
      print("Decrypted (Dart): $decryptedText");
      if (decryptedText == originalText) {
        print("Dart Encryption/Decryption Test: SUCCESSFUL");
      } else {
        print("Dart Encryption/Decryption Test: FAILED");
        print("Expected: $originalText, Got: $decryptedText");
      }
    } else {
      print("Dart Decryption FAILED (result was null).");
    }
  } else {
    print("Dart Encryption FAILED (result was null).");
  }

  print("\n--- Interoperability Test (Dart decrypts Python) ---");
  // Example: String "Hello from Python for Dart!" encrypted by Python
  // Python: StringCrypter().process_string('e', "Hello from Python for Dart!")
  final pythonEncrypted = "NzM2MzMxMzEzNjY0NjEzMDYxMzQzNjY0NjIzNjYyMzQzNFN1Z2RjY2RjZWNkZWNlY2RjZWRlY2VjZGNlZGVjZWNkY2VkZWNlY2RjZWRlY2VjZGNlZGVjZWNkY2VkZWNlY2Q="; // Placeholder
  final expectedPythonDecryption = "Hello from Python for Dart!"; // Placeholder

  print("Python Encrypted: $pythonEncrypted");
  final decryptedFromPython = processStringDart('d', pythonEncrypted);
  if (decryptedFromPython != null) {
    print("Decrypted by Dart: $decryptedFromPython");
    if (decryptedFromPython == expectedPythonDecryption) {
      print("Dart decryption of Python string: SUCCESSFUL (with correct Python output)");
    } else {
      print("Dart decryption of Python string: FAILED or placeholder data used.");
      print("Expected: $expectedPythonDecryption, Got: $decryptedFromPython");
    }
  } else {
    print("Dart decryption of Python string FAILED (result was null).");
  }
  print("Note: For the Python interop test to be meaningful, replace 'pythonEncrypted' with actual output from your Python script for the string '$expectedPythonDecryption'.");


  print("\n--- Interoperability Test (Dart encrypts for others) ---");
  final dartMsgForOthers = "Hello from Dart for other languages!";
  final dartEncryptedForOthers = processStringDart('e', dartMsgForOthers);
  if (dartEncryptedForOthers != null) {
    print("Dart Encrypted for others (Original: '$dartMsgForOthers'): $dartEncryptedForOthers");
    print("Take this string and try to decrypt it using process_string('d', ...) in other languages.");
  } else {
    print("Dart encryption for others FAILED (result was null).");
  }
}
